packages/react-storage/src/components/StorageBrowser/actions/handlers/listLocations.ts

```diff
@@ -1,21 +1,85 @@
-import { Permission } from '../../storage-internal';
+import {
+  ListLocationsOutput,
+  Permission,
+  listCallerAccessGrants,
+} from '../../storage-internal';
+import { assertListCallerAccessGrantInput } from '../../validators';
+
+import {
+  ListHandlerOptions,
+  ListHandlerInput,
+  ListHandlerOutput,
+  ListHandler,
+} from '../types';
 
-import { ListHandlerOptions, ListHandlerOutput, ListHandler } from '../types';
 import { LocationData, LocationType } from './types';
+import { parseLocations } from './utils';
+
+const DEFAULT_PAGE_SIZE = 1000;
 
 type ExcludeType = Permission | LocationType;
 
 export interface ListLocationsHandlerOptions
   extends ListHandlerOptions<ExcludeType | ExcludeType[]> {}
 
-export interface ListLocationsHandlerInput {
-  options?: ListLocationsHandlerOptions;
-}
+export interface ListLocationsHandlerInput
+  extends ListHandlerInput<ListLocationsHandlerOptions> {}
+
 export interface ListLocationsHandlerOutput
   extends ListHandlerOutput<LocationData> {}
 
 export interface ListLocationsHandler
   extends ListHandler<ListLocationsHandlerInput, ListLocationsHandlerOutput> {}
 
-export const listLocationsHandler: ListLocationsHandler =
-  null as unknown as ListLocationsHandler;
+const shouldExclude = (
+  permission: Permission,
+  type: LocationType,
+  exclude?: ExcludeType | ExcludeType[]
+) =>
+  !exclude
+    ? false
+    : typeof exclude === 'string'
+    ? exclude === permission || exclude === type
+    : exclude.includes(permission) || exclude.includes(type);
+
+export const listLocationsHandler: ListLocationsHandler = async (input) => {
```

Comment:
Where will the exhaustive while loop from the original listLocationAction go?

https://github.com/aws-amplify/amplify-ui/blob/feat-storage-browser/main/packages/react-storage/src/components/StorageBrowser/do-not-import-from-here/actions/listLocationsAction.ts#L76

This might be a question for @pranavosu or @calebpollman

Reply:
I think it goes in this new action. There's an opportunity to reuse some of the search logic, but that can be a follow up after fit and finish. Same for location detail handler.

Reply:
If you mean while loop to populate locations, it is recursion here right?

Reply:
Oh you're right, I didn't read it very well

Reply:
@reesscot if recursion does not read well i can update it `do..while`

Reply:
if we want to align to a certain pattern that makes sense to me but don't feel strongly here

Reply:
+1 If we can do it with while loop, I'd prefer it over tail recursion.

Reply:
Oh wait eat 🍕 Ashoka!